### PR TITLE
feat: Make Argo UI aware of OpenShift objects#4601

### LINKED ISSUE #4601: Make Argo UI aware of OpenShift objects, such as Route and DeploymentConfig
# Summary

In UI, when Argo renders a Route object box, it does not show the link to the URL it represents, as it happens to Ingress objects.
The same is true for a application box in Dashboard, which shows ingresses links, but not routes.
It seems to me that the problem is that Argo is not specifically aware of OpenShift objects.

# Motivation

Make Argo better to use when integrated to OpenShift clusters.

# Proposal

So far I had no synchronization problem with OpenShift, only UI problems. So I think that the way to go is to make the UI capable to distinguish some OpenShift objects, at least the most common ones, like DeploymentConfig and Route.

### PR DESCRIPTION
**Pull request by iam-veeramalla**
Signed-off-by: iam-veeramalla <abhishek.veeramalla@gmail.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

This PR will make Argo UI aware of OpenShift objects. It provides details such as the route url and services attached to route. This feature is currently available for ingress.

Fixes #4601 #7052

### CODE REVIEW

**Comment by iam-veeramalla** [comment]
@alexmt can you review this ?

**Comment by jessesuen** [comment]
I think @jgwest and/or @wtam2018 would be good ones to review this

**Comment by wtam2018** [comment]
@iam-veeramalla do you mind to create an issue to describe the enhancement?  It would be nice to include some screen shots to  illustrate the enhancement.

**Comment by iam-veeramalla** [comment]
> @iam-veeramalla do you mind to create an issue to describe the enhancement? It would be nice to include some screen shots to illustrate the enhancement.


Sure @wtam2018

**Comment by iam-veeramalla** [comment]
> > @iam-veeramalla do you mind to create an issue to describe the enhancement? It would be nice to include some screen shots to illustrate the enhancement.
> 
> Sure @wtam2018

Argo renders the Ingress url for Ingress resource deployed in k8s, This will help users to access their applications(pods) directly from the Argo UI. This feature is not available for openshift routes as Argo does not identify/render/provide any route details. Also, by the nature of openshift, it will convert any ingress resource into routes. So, Argo does not render the url, health details of k8s ingress or openshift routes deployed in openshift. 

This enhancement will allow users to access the openshift route url from Argo UI. This will make user experience better as they don't have to describe route using `oc cli` and frame the route url(typically `host+path`) to access their applications. This feature is already available for ingress and istio virtual service. This feature is added as per the requirement in issue  https://github.com/argoproj/argo-cd/issues/4601

Screenshot 2020-12-22 at 4 29 21 PM


Another screenshot to explain how this enhancement will provide 
Screenshot 2020-12-22 at 4 50 05 PM

Openshift route object identified by Argo does not even provide the health details. I will create an issue for that and add it as a follow up enhancement to this.

Please let me know, If you need more details.

**Comment by wtam2018** [comment]
Fixes #4601

**Comment by iam-veeramalla** [comment]
I think this is good to merge @jessesuen

**Comment by wtam2018** [comment]
I'll take a look.

File: controller/cache/info.go [hunk 1]
```diff
@@ -170,6 +191,45 @@ func populateIngressInfo(un *unstructured.Unstructured, res *ResourceInfo) {
...
+	route := getRoute(un)
+	targetsMap := make(map[v1alpha1.ResourceRef]bool)
+
+	if backend, ok, err := unstructured.NestedMap(un.Object, "spec", "to"); ok && err == nil {
+		targetsMap[v1alpha1.ResourceRef{
+			Kind:      kube.ServiceKind,
+			Namespace: un.GetNamespace(),
+			Name:      fmt.Sprintf("%s", backend["name"]),
+		}] = true
+	}
```
Comment (wtam2018) [inline controller/cache/info.go:204]: There is only one `to` backend service per route but there can be a list of `alternateBackends`.  Please see https://docs.openshift.com/container-platform/3.6/architecture/networking/routes.html#alternateBackends

File: controller/cache/info.go [hunk 2]
```diff
@@ -170,6 +191,45 @@ func populateIngressInfo(un *unstructured.Unstructured, res *ResourceInfo) {
 	res.NetworkingInfo = &v1alpha1.ResourceNetworkingInfo{TargetRefs: targets, Ingress: ingress, ExternalURLs: urls}
 }
 
+func populateRouteInfo(un *unstructured.Unstructured, res *ResourceInfo) {
+	route := getRoute(un)
+	targetsMap := make(map[v1alpha1.ResourceRef]bool)
+
+	if backend, ok, err := unstructured.NestedMap(un.Object, "spec", "to"); ok && err == nil {
+		targetsMap[v1alpha1.ResourceRef{
+			Kind:      kube.ServiceKind,
```
Comment (wtam2018) [inline controller/cache/info.go:200]: Currently, `Kind` can only be `Service` currently but should we not hardcode here?   Perhaps, use `backend["kind"]` instead?

File: controller/cache/info.go [hunk 3]
```diff
@@ -74,6 +79,22 @@ func getIngress(un *unstructured.Unstructured) []v1.LoadBalancerIngress {
...
 
+func getRoute(un *unstructured.Unstructured) []v1.LoadBalancerIngress {
+	route, ok, err := unstructured.NestedSlice(un.Object, "status", "ingress")
+	if !ok || err != nil {
+		return nil
+	}
+	res := make([]v1.LoadBalancerIngress, 0)
+	for _, item := range route {
+		if lbIngress, ok := item.(map[string]interface{}); ok {
+			if hostname := lbIngress["routerCanonicalHostname"]; hostname != nil {
```
Comment (wtam2018) [inline controller/cache/info.go:90]: `routerCanonicalHostname` is optional.   Please see https://docs.openshift.com/container-platform/4.6/rest_api/network_apis/route-route-openshift-io-v1.html

File: controller/cache/info.go [hunk 4]
```diff
@@ -170,6 +191,45 @@ func populateIngressInfo(un *unstructured.Unstructured, res *ResourceInfo) {
...
+			path = strings.TrimSuffix(path, "*")
+			url += path
+		}
+		if res.NetworkingInfo != nil {
+			urls = res.NetworkingInfo.ExternalURLs
+		}
+		urls = append(urls, url)
+	}
+
+	res.NetworkingInfo = &v1alpha1.ResourceNetworkingInfo{TargetRefs: targets, Ingress: route, ExternalURLs: urls}
```
Comment (wtam2018) [inline controller/cache/info.go:229]: I think we need to handle the edge cases when we are unable to extract backend services, ingress, and urls.

**Review by wtam2018 — CHANGES_REQUESTED** [review]
#4601 also propose to support  `DeploymentConfig` .     Should we address  it?    We could do it in a different PR.

**Comment by dlmorais-pbh** [comment]
> #4601 also propose to support `DeploymentConfig` . Should we address it? We could do it in a different PR.

One thing I know that Argo supports for Deployment, but not for DeploymentConfig, is the ability to trigger a rollout by using the "three dots menu" on Deployment object.

And there is also the icon at object's box.

**Comment by mbecca** [comment]
any update for this PR?

**Comment by iam-veeramalla** [comment]
> any update for this PR?

I started to work back on this. I will probably get a PR for this by end of the week :)

**Comment by pasha-codefresh** [comment]
@iam-veeramalla could you please update on this? Do you need some help?

**Comment by crenshaw-dev** [comment]
Closing for now, @iam-veeramalla feel free to reopen when ready. :-)

**Pull request closed without merging**